update core IBC docs (#7560)
cosmos/cosmos-sdk: x/ibc/core/spec/01_concepts.md

> NOTE: if you are not familiar with the IBC terminology and concepts, please read
this [document](https://github.com/cosmos/ics/blob/master/ibc/1_IBC_TERMINOLOGY.md) as prerequisite reading.

## Client Creation, Updates, and Upgrades

IBC clients are on chain light clients. The light client is responsible for verifying
counterparty state. A light client can be created by any user submitting a client
identifier and a valid initial `ClientState` and `ConsensusState`. The client identifier
must not already be used. Clients are given a client identifier prefixed store to
store their associated client state and consensus states. Consensus states are 
stored using their associated height. 

> Clients can be updated by any user submitting a valid `Header`. The client state callback

Reviewer: I think it's important to mention that the update is not necessarily the latest height.

to `CheckHeaderAndUpdateState` is responsible for verifying the header against previously
stored state. The function should also return the updated client state and consensus state 
if the header is considered a valid update. A light client, such as Tendermint, may have
client specific parameters like `TrustLevel` which must be considered valid in relation
to the `Header`.

> Clients may be upgraded. The upgrade should be verified using `VerifyUpgrade`. It is not

Reviewer: I think it's also relevant to mention how to upgrade the client if one of the chains halts/panics vs if the upgrade is planned

Author: This should go at client specific docs IMO since it could be implemented in a multitude of ways depending on the client

Reviewer: ah yes. Although I think it's still relevant to add that as a comment (i.e that it is up to the light client implementation).

a requirement to allow for light client upgrades. For example, the solo machine client 
will simply return an error on `VerifyUpgrade`.

## Client Misbehaviour

> IBC clients must freeze when the counterparty chain becomes malicious and 

Reviewer: ```suggestion
IBC clients must freeze when the counterparty chain becomes byzantine and 
```

takes actions that could fool the light client into accepting invalid state 
transitions. Thus, relayers are able to submit Misbehaviour proofs that prove 
that a counterparty chain has signed two Headers for the same height. This 
constitutes misbehaviour as the IBC client could have accepted either header 
as valid. Upon verifying the misbehaviour the IBC client must freeze at that 
height so that any proof verifications for the frozen height or later fail.

Note, there is a difference between the chain-level Misbehaviour that IBC is 
concerned with and the validator-level Evidence that Tendermint is concerned 

> with. Tendermint must be able to detect, submit, and punish any evidence of 

Reviewer: I'd prob include a reference to the tendermint evidence docs here

Author: this was existing docs (just got moved around). @AdityaSripal would know better than I on what TM evidence docs to link

Implementations which do not feel they would benefit from versioning can do
basic string matching using a single compatible version.

## Sending, Receiving, Acknowledging Packets

> A packet may be associated with one of the following states:

Reviewer: I think it's important to mention that the packet can be received but the application must still know how to unmarshal the packet data bytes to the concrete types in order to execute the business logic successfully.

Author: I suppose so, but this is mostly docs for the `core` code, not really the entire IBC stack. To the `core`, the packet data bytes don't mean anything. This would be better to put under the `docs` imo

- the packet does not exist (ie it has not been sent)
- the packet has been sent but not received (the packet commitment exists on the 
sending chain, but no receipt exists on the receiving chain)
- the packet has been received but not acknowledged (packet commitment exists 
on the sending chain, a receipt exists on the receiving chain, but no acknowledgement
exists on the receiving chain)
- the packet has been acknowledgement but the acknowledgement has not been relayed 
(the packet commitment exists on the sending chain, the receipt and acknowledgement
exist on the receiving chain)

> - the packet has completed its life cycle (the packet commitment does not exist on

Reviewer: nit, I think we should clarify the difference between a receipt and an acknowledgment

Author: added a small terminology section before this

Each packet is required to have at least one valid timeout field. 

## Closing Channels

Closing a channel occurs in occurs in 2 handshake steps as defined in [ICS 04](https://github.com/cosmos/ics/tree/master/spec/ics-004-channel-and-packet-semantics).

`ChanCloseInit` will close a channel on the executing chain if the channel exists, it is not 
already closed and the connection it exists upon is OPEN. Channels can only be closed by a 
calling module or in the case of a packet timeout on an ORDERED channel.

`ChanCloseConfirm` is a response to a counterparty channel executing `ChanCloseInit`. The channel
on the executing chain will be closed if the channel exists, the channel is not already closed, 
the connection the channel exists upon is OPEN and the executing chain successfully verifies
that the counterparty channel has been closed.

## Port and Channel Capabilities

## Hostname Validation

The 24-host sub-module parses and validates identifiers. It also builds 
the key paths used to store IBC related information. 

A valid identifier must conatin only alphanumeric characters or the 
following list of allowed characters: 
".", "\_", "+", "-", "#", "[", "]", "<", ">" 

> Client identifiers must contain between 9 and 64 characters.
> Connection identifiers must contain between 10 and 64 characters.
> Channel identifiers must contain between 10 and 64 characters.
> Port identifiers must contain between 2 and 64 characters.

Reviewer: ```suggestion
- Client identifiers must contain between 9 and 64 characters.
- Connection identifiers must contain between 10 and 64 characters.
- Channel identifiers must contain between 10 and 64 characters.
- Port identifiers must contain between 2 and 64 characters.
```

on the sending chain, a receipt exists on the receiving chain, but no acknowledgement
exists on the receiving chain)
- the packet has been acknowledgement but the acknowledgement has not been relayed 
(the packet commitment exists on the sending chain, the receipt and acknowledgement
exist on the receiving chain)
- the packet has completed its life cycle (the packet commitment does not exist on
the sending chain, but a receipt and acknowledgement exist on the receiving chain)

Sending of a packet is initiated by a call to the `ChannelKeeper.SendPacket` 
function by an application module. Packets being sent will be verified for
correctness (core logic only). If the packet is valid, a hash of the packet
will be stored as a packet commitment using the packet sequence in the key.
Packet commitments are stored on the sending chain.

A message should be sent to the receving chain indicating that the packet
has been committed on the sending chain and should be received on the 
receiving chain. The light client on the receiving chain, which verifies
the sending chain's state, should be updated to the lastest sending chain

> state if possible. The verification will fail if the lastest state of the

Reviewer 2: ```suggestion
state if possible. The verification will fail if the latest state of the
```

light client does not include the packet commitment. The receiving chain
is responsible for verifying that the counterparty set the hash of the 
packet. If verification of the packet to be received is successful, the
receiving chain should store a receipt of the packet and call application

> logic if necessary. An acknowledgement may be stored at this time (synchonously)

Reviewer 2: ```suggestion
logic if necessary. An acknowledgement may be processed and stored at this time (synchronously)
```

> or at another point in the future (aynchronously). 

Reviewer 2: ```suggestion
or at another point in the future (asynchronously). 
```

Acknowledgements written on the receiving chain may be verified on the 
sending chain. If the sending chain successfully verifies the acknowledgement
then it may delete the packet commitment stored at that sequence. There is
no requirement for acknowledgements to be written. Only the hash of the
acknowledgement is stored on the chain. Application logic may be executed
in conjunction with verifying an acknowledgement. For example, in fungible
cross-chain token transfer, a failed acknowledgement results in locked or
burned funds being refunded. 

Relayers are responsible for reconstructing packets between the sending, 
receiving, and acknowledging of packets. 

IBC applications sending and receiving packets are expected to appropriately
handle data contained within a packet. For example, cross-chain token 
transfers will unmarshal the data into proto definitions representing
a token transfer. 

> Future optimizations may allow for storage cleanup of stored packet 

Reviewer 2: Packet commitments are cleaned-up when acknowledgements are processed

Reviewer 2: (you mention this above)

Author: reworded

commitments and acknowledgements that no longer provide any usefulness. 
This may be from packets that have completed their life cycles or from
channels which do not require written acknowledgements. 

## Timing out Packets

> A packet may be timed out on the receiving chain if the packet timeouts have

Reviewer 2: ```suggestion
A packet may be timed out on the receiving chain if the packet timeout height or timestamp has
```

surpassed on the receiving chain for a timeout to be valid. A timeout timestamp 
or timeout height with a 0 value indicates the timeout field may be ignored. 
Each packet is required to have at least one valid timeout field. 

## Closing Channels

Closing a channel occurs in occurs in 2 handshake steps as defined in [ICS 04](https://github.com/cosmos/ics/tree/master/spec/ics-004-channel-and-packet-semantics).

`ChanCloseInit` will close a channel on the executing chain if the channel exists, it is not 
already closed and the connection it exists upon is OPEN. Channels can only be closed by a 
calling module or in the case of a packet timeout on an ORDERED channel.

`ChanCloseConfirm` is a response to a counterparty channel executing `ChanCloseInit`. The channel
on the executing chain will be closed if the channel exists, the channel is not already closed, 
the connection the channel exists upon is OPEN and the executing chain successfully verifies
that the counterparty channel has been closed.

## Port and Channel Capabilities

## Hostname Validation

The 24-host sub-module parses and validates identifiers. It also builds 
the key paths used to store IBC related information. 

> A valid identifier must conatin only alphanumeric characters or the 

Reviewer 2: Reference ICS 24, which should match this
